data/transactions/logic/eval.go

```diff
@@ -0,0 +1,1193 @@
+		cx.err = fmt.Errorf("program version %d greater than max supported version %d", version, EvalMaxVersion)
+		return false, cx.err
+	}
+	if (params.Proto != nil) && (version > params.Proto.LogicSigVersion) {
+		cx.err = fmt.Errorf("program version %d greater than protocol supported version %d", version, params.Proto.LogicSigVersion)
+		return false, cx.err
+	}
+	// TODO: if EvalMaxVersion > version, ensure that inaccessible
+	// fields as of the program's version are zero or other
+	// default value so that no one is hiding unexpected
+	// operations from an old program.
+	cx.version = version
+	cx.pc = vlen
+	cx.EvalParams = params
+	cx.stack = make([]stackValue, 0, 10)
+	cx.program = program
+	for (cx.err == nil) && (cx.pc < len(cx.program)) {
+		cx.step()
+		cx.stepCount++
+		if cx.stepCount > len(cx.program) {
+			return false, errLoopDetected
+		}
+		if params.Proto != nil && uint64(cx.cost) > params.Proto.LogicSigMaxCost {
+			return false, errCostTooHigh
+		}
+	}
+	if cx.err != nil {
+		if cx.Trace != nil {
+			fmt.Fprintf(cx.Trace, "%3d %s\n", cx.pc, cx.err)
+		}
+		return false, cx.err
+	}
+	if len(cx.stack) != 1 {
+		if cx.Trace != nil {
+			fmt.Fprintf(cx.Trace, "end stack:\n")
+			for i, sv := range cx.stack {
+				fmt.Fprintf(cx.Trace, "[%d] %s\n", i, sv.String())
+			}
+		}
+		return false, fmt.Errorf("stack len is %d instead of 1", len(cx.stack))
+	}
+	if cx.stack[0].Bytes != nil {
+		return false, errors.New("stack finished with bytes not int")
+	}
+	return cx.stack[0].Uint != 0, nil
+}
+
+// Check should be faster than Eval.
+// Returns 'cost' which is an estimate of relative execution time.
+func Check(program []byte, params EvalParams) (cost int, err error) {
+	defer func() {
+		if x := recover(); x != nil {
+			buf := make([]byte, 16*1024)
+			stlen := runtime.Stack(buf, false)
+			cost = 0
+			errstr := string(buf[:stlen])
+			if params.Trace != nil {
+				if sb, ok := params.Trace.(*strings.Builder); ok {
+					errstr += sb.String()
+				}
+			}
+			err = PanicError{x, errstr}
+		}
+	}()
+	if (params.Proto != nil) && (params.Proto.LogicSigVersion == 0) {
+		err = errLogicSignNotSupported
+		return
+	}
+	var cx evalContext
+	version, vlen := binary.Uvarint(program)
+	if vlen <= 0 {
+		cx.err = errors.New("invalid version")
+		return 0, cx.err
+	}
+	if version > EvalMaxVersion {
+		err = fmt.Errorf("program version %d greater than max supported version %d", version, EvalMaxVersion)
+		return
+	}
+	if (params.Proto != nil) && (version > params.Proto.LogicSigVersion) {
+		err = fmt.Errorf("program version %d greater than protocol supported version %d", version, params.Proto.LogicSigVersion)
+		return
+	}
+	cx.version = version
+	cx.pc = vlen
+	cx.EvalParams = params
+	cx.checkStack = make([]StackType, 0, 10)
+	cx.program = program
+	for (cx.err == nil) && (cx.pc < len(cx.program)) {
+		cost += cx.checkStep()
+	}
+	if cx.err != nil {
+		err = fmt.Errorf("%3d %s", cx.pc, cx.err)
+		return
+	}
+	if len(cx.checkStack) != 1 {
```

Comment:
Following discussion with @pzbitskiy:

One counterintuitive behavior of `Check` is that because it doesn't account for branches, it rejects programs that are valid when executed, but which leaves the same number of things on the stack in both sides of the branch and are thus rejected as they leave too many things on the stack. 

For instance, consider the following statement in a hypothetical high-level language:
```
if <cond>:
   x = <then-value>
otherwise:
   x = <else-value>
return x > 5
```

A natural way to compile this would be into the following assembly:
```
<cond>
bnz then
<else-condition>
int 1
bnz end
then:
<then-condition>
end:
5
>
```

Even though this program should execute correctly, because `Check` does not know that both branches are mutually exclusive, it will reject this program as leaving too much data on the stack.